### PR TITLE
Add pure julia exp10 function

### PR DESCRIPTION
**Pull request by musm**
Need ~~to add tests here and possibly benchmarks~~ https://github.com/JuliaCI/BaseBenchmarks.jl/pull/73. 

```julia
julia> @benchmark Base.exp10(5.0)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     105.795 ns (0.00% GC)
  median time:      155.140 ns (0.00% GC)
  mean time:        170.898 ns (0.00% GC)
  maximum time:     4.148 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000

julia> @benchmark My.exp10(5.0)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     13.816 ns (0.00% GC)
  median time:      15.395 ns (0.00% GC)
  mean time:        16.895 ns (0.00% GC)
  maximum time:     252.250 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000


test range -10:0.00021:10, -35:0.00023:1000, -300:0.001:300

Accuracy tests for Float64
     exp10          : max 0.91517196 at x = 275.61822                : mean 0.10907410
Base.exp10          : max 0.83238916 at x = 269.27436                : mean 0.11107440

Accuracy tests for Float32
     exp10          : max 0.87275678 at x = -3.13534                   : mean 0.02482310
Base.exp10          : max 0.82333934 at x = -1.05200                   : mean 0.02529399
```

It is expected that base will be slightly more accurate since it is just calling `^` function

### CODE REVIEW

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by musm** [comment]
I haven't (yet) added any specific benchmarks for exp10 and base benchmarks doesn't have any benchmarks that include exp10 ...

File: base/special/exp10.jl [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+#    sup norm error within the interval specified, of the polynomial approximations:
+#    Float64 : [2.7245504724394698951963418186308011736517648015976e-18; 2.72455298957534767204462079073786522870226850867e-18]
+#    Float32 : [9.602647147784220587121706280697708280058577656746e-10; 9.6026560194009888671163788405817409356838220899077e-10]
+#
+# 3. Scale back: exp10(x) = 2^k * exp10(r)
+
+# log2(10)
+const LOG210 = 3.321928094887362347870319429489390175864831393024580612054756395815934776608624
+# log10(2)
+const LOG102 = 3.010299956639811952137388947244930267681898814621085413104274611271081892744238e-01
```
Comment (StefanKarpinski) [inline base/special/exp10.jl]: Maybe have underscores between the numbers, i.e. `LOG2_10` and `LOG10_2`?

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21445#issuecomment-295415082) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d6929d5ca118d3cb30a2cf00db02209e79d8cde5/2ad05cd_vs_e6f8180/report.md). cc @jrevels

**Comment by simonbyrne** [comment]
Looks good in general.

Do you have any accuracy results? I tested the `Float32`s and I noticed that some gave an error > 1ulp, e.g. `-0.000122065285f0`. The cause in that case seemed to be that the branch to the Taylor approximation. `exp10_small_thres` may need to be a bit smaller.

**Comment by musm** [comment]
Please see
https://github.com/JuliaLang/julia/pull/21445/files#diff-2b4b439b810120cb21b0b4eed48a360aR11
which is indeed <1ulp, but outside this range it may be slightly larger than 1ulp, I think I have observed numbers 1.1 or 1.2 ulp at most in the past but haven't hit one yet...



and in the original post
```julia
test range -10:0.00021:10, -35:0.00023:1000, -300:0.001:300

Accuracy tests for Float64
     exp10          : max 0.91517196 at x = 275.61822                : mean 0.10907410
Base.exp10          : max 0.83238916 at x = 269.27436                : mean 0.11107440

Accuracy tests for Float32
     exp10          : max 0.87275678 at x = -3.13534                   : mean 0.02482310
Base.exp10          : max 0.82333934 at x = -1.05200                   : mean 0.02529399
```


> Do you have any accuracy results? I tested the Float32s and I noticed that some gave an error > 1ulp, e.g. -0.000122065285f0. The cause in that case seemed to be the branch to the Taylor approximation. exp10_small_thres may need to be a bit smaller.

Unlike the exp function,   I can find   small number for float32 where it's not 1ulp. So we can either get rid of the small Taylor branch or keep it.

**Comment by simonbyrne** [comment]
As I said above, I think the main issue is simply that the threshold is too high: the next term in the Taylor series is `(log(10)*x)^2/2`, and so if you choose a value at the threshold, then:
```
julia> x = Base.Math.exp10_small_thres(Float32)
0.00012207031f0

julia> (x*log(10f0))^2/2 / eps(0.9f0)
0.6627373f0

i.e. 0.66 ulps if `x` is negative (so `exp10(x) < 1`). Combined the potential error in evaluating `1 + x*log(10f0)` (which due to the size of the other arguments, is dominated by the `+`, so about ~0.5 ulps), and that will be enough to push you over the 1 ulp.

If you choose the threshold to be half the current value, then you should be okay.

File: base/special/exp10.jl [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    2.302585124969482421875f0,
+    2.650949001312255859375f0,
+    2.0346698760986328125f0,
+    1.17125606536865234375f0,
+    0.5400512218475341796875f0,
+    0.20749187469482421875f0,
+    5.2789829671382904052734375f-2)
+
+@eval exp10_small_thres(::Type{Float64}) = $(2.0^-29)
+@eval exp10_small_thres(::Type{Float32}) = $(2.0f0^-13)
```
Comment (simonbyrne) [inline base/special/exp10.jl]: Changing this to be `2.0f0^-14` gave a max ulp error of `0.9558756425976753` for all `Float32` values.

File: base/special/exp10.jl (hunk 1 above)
Comment (vtjnash) [inline base/special/exp10.jl]: These (and the horner macro) seem like suspiciously many digits, when they will have to get truncated internally when they are read in (e.g. this one will become 0.3010299956639812). Is it typical to over-state the accuracy in this way?

File: base/special/exp10.jl (hunk 2 above)
Comment (musm) [inline base/special/exp10.jl]: Are you sure about this? How are you testing over all float32, can you post the script. For example I'm finding many numbers greater than 1 ulp. The max float32 ulp I have found is at least 1.158, yet where these occur are larger than the small threshold value, even the value you post (-0.000122065285f0) is larger than the threshold value

Here is my test script to double check

```julia
function countulp(::Type{T}, x::AbstractFloat, y::AbstractFloat) where T
    isnan(x) && isnan(y) && return zero(T)
    isnan(x) || isnan(y) && return T(1000)
    if isinf(T(x)) && isinf(T(y))
        return signbit(x) == signbit(y) ? zero(T) : T(1001)
    end
    if x == zero(x) && y == zero(y)
        return signbit(x) == signbit(y) ? zero(T) : T(1002)
    end
    if isfinite(x) && isfinite(y)
        return T(abs(x - y)/eps(T(y)))
    end
    return T(1003)
end

start32 = 0x00000000
end32 = 0xffffffff

function test_acc(f1,f2,::Type{T}, suint::U, euint::U, spacing::Integer=1000) where T where U<:Unsigned
    spacing_uint = convert(typeof(suint), spacing)
    vuint = suint
    x = reinterpret(T, vuint)
    ulp = zero(T)
    while vuint < euint
        v1 =f1(x)
        v2 = f2(big(x))
        ulp_ = countulp(T,v1,v2)
        ulp = max(ulp,ulp_)
        if ulp_ > T(1.0)
            @show x,ulp
        end
        vuint += spacing_uint
        x = reinterpret(T, vuint)
    end
    ulp
end

@time ulp,x = test_acc(Amal.exp10, Base.exp10, Float32, start32, end32)
```

Comment (simonbyrne) [inline base/special/exp10.jl]: I just checked against `Float64`: I used the following:
```
function check32()
    maxulps = 0.0
    xmax    = 0f0
    for u in typemin(UInt32):typemax(UInt32)
        x = reinterpret(Float32, u)
        isfinite(x) || continue
        
        y32 = exp10(x)
        y64 = exp10(Float64(x))
        if !isfinite(Float32(y64))
            continue
        end
        if !isfinite(y32)
            error("Non-finite value at $x")
        end
        d = abs(y32-y64)/eps(Float32(y64))
        if d > maxulps
            maxulps = d
            xmax = x
        end
    end
    maxulps,xmax
end
```

which gives:
```
julia> check32()
(0.9558756425976753, -3.7376504f0)
```

**Comment by musm** [comment]
updated, can someone trigger nanosoldier? thanks

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21445#issuecomment-297966296) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8f0564273358e840d7eb76b3ebfe54463a894ebc/bff6d1d_vs_991b581/report.md). cc @jrevels

**Comment by musm** [comment]
benchmarks look good

**Comment by tkelman** [comment]
The manually written-out coefficients look considerably longer than they need to be, as a Float64 aren't many of the trailing digits not accomplishing anything?

**Comment by musm** [comment]
the constants are written out to higher precision than needed since they can be useful for implementing higher precision function if/when float128 support is added in julia. I'd rather not change that and it is harmless anyways.

**Comment by musm** [comment]
As a reminder this should not be squashed when merged

**Comment by tkelman** [comment]
Does the first commit pass on its own?

**Comment by musm** [comment]
Yes. The second commit fixes the fast_math variants to actually call the pure julia exp and exp10 functions

**Comment by tkelman** [comment]
@simonbyrne go ahead and merge if you approve of this - though it's not the end of the world if we hold off until branching on this one

**Comment by musm** [comment]
More data points,  I tested 2^30 \approx 1 billion random float64 values   (with the help of a server and julia sharedarrays)  the max error in ulp I found is 1.2623 at x = -210.2442424463067
and 0.00893% of the random values had an ulp greater than 1.0

**Comment by ViralBShah** [comment]
This is great. I do think it is best to hold off until branching, which I suspect is this week.

File: base/special/exp10.jl [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    if xa > reinterpret(Unsigned, EXP10_MAX(T))
+        if xa >= exponent_mask(T)
+            xa & significand_mask(T) != 0 && return T(NaN)
+            return xsb ? T(0.0) : T(Inf) # exp10(+-Inf)
+        end
+        x > EXP10_MAX(T) && return T(Inf)
+        x < EXP10_MIN(T) && return T(0.0)
+    end
+    # compute approximation
+    if xa > reinterpret(Unsigned, T(0.5)*T(LOG10_2)) # |r| > 0.5 log10(2).
```
Comment (simonbyrne) [inline base/special/exp10.jl]: maybe change the comment to be `|x|` instead of `|r|`

**Comment by simonbyrne** [comment]
I've noticed that the accuracy can be quite a bit lower on machines without native FMA (or those using the prebuilt images). On such machines I see a lot of values with > 1 ulp error.

It is sometimes possible to avoid this by rearranging the last few operations. However I don't have time to look at it at the moment, so I think we should merge this and do some more experiments before the next release.

**Comment by StefanKarpinski** [comment]
Please do open an issue about investigating the non-native FMA thing since otherwise we'll forget.

**Comment by musm** [comment]
Yes without fma it is *very* difficult and perhaps not possible (unless you use double/double floats which is kinda not desirable). If I do play around with this,  I recall trying the same tricks as in exp by construction a rational  but I don't think that worked very well at all (I don't even have a copy of this test; must have deleted probably because the accuracy wasn't good enough).

**Comment by simonbyrne** [comment]
For future reference, one such value was `x = -1.6515175124292971`, which gives 1.06 ulps error.

**Comment by musm** [comment]
Did u see my post which was testing hardware fma

> More data points, I tested 2^30 \approx 1 billion random float64 values (with the help of a server and julia sharedarrays) the max error in ulp I found is 1.2623 at x = -210.2442424463067
and 0.00893% of the random values had an ulp greater than 1.0

**Comment by simonbyrne** [comment]
Ah, I forgot about that. Good point (for the record, I only see that without FMA, with FMA is ~0.74 ulps)

**Comment by musm** [comment]
That was with FMA testing against BigFloats which is *very* expensive.  I had to use a hefty server and the computation ran for at least 1 hour.

**Comment by pkofod** [comment]
Would you mind posting the script?

**Comment by simonbyrne** [comment]
My script was here:
https://gist.github.com/simonbyrne/6dcfac9488a44f1fb2a64b82c01e88cb

**Comment by musm** [comment]
thoughts?

**Comment by musm** [comment]
BTW I think it's fine for now that this function is probably at most 1.5 ulps for Float64 (probably closer to 1.3ulp...) . Not many libms even have a exp10 implementation,  e.g intel's math libs don't....

I don't think you can get much better with a polynomial in this case.

File: base/special/exp.jl [hunk 4]
```diff
@@ -32,22 +32,22 @@
...
 LN2U(::Type{Float32}) = 6.9313812256f-1
 
 LN2L(::Type{Float64}) = 1.90821492927058770002e-10
 LN2L(::Type{Float32}) = 9.0580006145f-6
 
-# max and min arguments for exponential fucntions
-MAXEXP(::Type{Float64}) = 7.09782712893383996732e2 # log 2^1023*(2-2^-52)
-MAXEXP(::Type{Float32}) = 88.72283905206835f0      # log 2^127 *(2-2^-23)
+# max and min arguments
+EXP_MAX(::Type{Float64}) = 7.09782712893383996732e2 # log 2^1023*(2-2^-52)
```
Comment (musm) [inline base/special/exp.jl]: bikeshedding but do people prefer
`EXP_MAX`
`MAX_EXP`
or the no underscore variants of the above

Comment (StefanKarpinski) [inline base/special/exp.jl]: `MAX_EXP`

Comment (musm) [inline base/special/exp.jl]: Ok Thanks I have changed it

**Comment by musm** [comment]
We are post 0.6, would it be possible to decide whether this is wanted or not?

**Comment by KristofferC** [comment]
Planning on merging this tomorrow unless I hear any objections

**Comment by KristofferC** [comment]
Thanks for another cool PR, showing what Julia is capable of @musm

**Comment by musm** [comment]
wow nice 

So this function took me a long time to write. I'm still slightly disappointed I couldn't get <1 ulp on the Float64 version. Not sure on @pkofod plans but since he is working on the libm this summer, I'll say that it is possible to get  0.5 ulp accuracy and (probably) beat the speed of the polynomial version at least on scalars via a table method, if that is something he is interested to pursue. (I looked at this but don't have anything so polished to be a PR and can't promise I will spend more time on it)

**Pull request merged**